x/src/lint.rs

```diff
@@ -0,0 +1,164 @@
+use crate::{config::Config, utils::project_root};
+use anyhow::anyhow;
+use std::{
+    fs,
+    path::{Path, PathBuf},
+    process::{Command, Stdio},
+    str,
+};
+use structopt::StructOpt;
+
+#[derive(Debug, StructOpt)]
+pub struct Args {
+    #[structopt(long)]
+    fail_fast: bool,
+}
+
+pub fn run(args: Args, _config: Config) -> crate::Result<()> {
+    let mut errors = false;
+    let project_root = project_root();
+    let mut full_file_path = PathBuf::new();
+
+    for file in get_file_listing()? {
+        full_file_path.push(&project_root);
+        full_file_path.push(&file);
+
+        let contents = match fs::read_to_string(&full_file_path) {
+            Ok(s) => s,
+            // Skip non utf8 files
+            Err(_) => continue,
+        };
+
+        // Run each lint on the file
+        for lint in LINTS {
+            if let Err(e) = lint(&file, &contents) {
+                errors = true;
+                let msg = format!("{}: {}", file.display(), e);
+                if args.fail_fast {
+                    return Err(anyhow!(msg));
+                } else {
+                    println!("{}", msg);
+                }
+            }
+        }
+    }
+
+    if errors {
+        Err(anyhow!("there were lint errors"))
+    } else {
+        Ok(())
+    }
+}
+
+static LINTS: &[fn(&Path, &str) -> Result<(), String>] = &[
+    has_license_header,
+    has_newline_at_eof,
+    has_trailing_whitespace,
+];
+
+fn has_newline_at_eof(_file: &Path, contents: &str) -> Result<(), String> {
+    if !contents.ends_with('\n') {
+        Err(format!("missing a newline at EOF"))
+    } else {
+        Ok(())
+    }
+}
+
+fn has_trailing_whitespace(_file: &Path, contents: &str) -> Result<(), String> {
+    for (ln, line) in contents
+        .lines()
+        .enumerate()
+        .map(|(ln, line)| (ln + 1, line))
+    {
+        if line.trim_end() != line {
+            return Err(format!("trailing whitespace on line {}", ln));
+        }
+    }
+
+    if contents
+        .lines()
+        .rev()
+        .take_while(|line| line.is_empty())
+        .count()
+        > 0
+    {
+        return Err(format!("trailing whitespace at EOF"));
+    }
+
+    Ok(())
+}
+
+static LICENSE_HEADER: &str = "Copyright (c) The Libra Core Contributors\n\
+                               SPDX-License-Identifier: Apache-2.0\n\
+                               ";
+
+fn has_license_header(file: &Path, contents: &str) -> Result<(), String> {
+    enum FileType {
+        Rust,
+        Shell,
+        MoveIr,
+        Proto,
+    }
```

Comment:
why not

```
enum FileType {
    Rust = "rs",
...
}
```

Reply:
Perhaps we still need the boiler plate code anyway... just curious to hear your thoughts.

Reply:
you can't do this, only integer types are accepted with that style of enum.